File: tests/mgcv_reference/generate.py
# SPDX-License-Identifier: GPL-3.0-or-later
# Copyright (C) 2026 Johannes Brachem

"""Regenerate committed mgcv oracle assets without adding R to pytest.

Run this file directly. It installs the requested local mgcv source into a temporary
R library, invokes the adjacent R generator, and writes compressed NumPy assets plus a
checksum manifest.
"""

from __future__ import annotations

import argparse
import hashlib
import json
import os
import shutil
import subprocess
import tempfile
from pathlib import Path

import numpy as np

HERE = Path(__file__).resolve().parent


def _matrix(path: Path) -> np.ndarray:
    return np.loadtxt(path, delimiter=",", ndmin=2)


def _metadata(path: Path) -> dict[str, int]:
    result: dict[str, int] = {}
    for line in path.read_text().splitlines():
        key, value = line.split()
        result[key] = int(value)
    return result


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--mgcv-source", type=Path, required=True)
    parser.add_argument("--output", type=Path, default=HERE / "assets")
    args = parser.parse_args()
    source = args.mgcv_source.resolve()
    output = args.output.resolve()
    output.mkdir(parents=True, exist_ok=True)

    with tempfile.TemporaryDirectory(prefix="smoothcon-mgcv-") as temporary:
        root = Path(temporary)
        library = root / "library"
        source_copy = root / "mgcv"
        csv_output = root / "csv"
        library.mkdir()
        shutil.copytree(source, source_copy)
        # Some macOS R installations define ENABLE_NLS without shipping the gettext
        # development headers. NLS only affects translated error messages and is not
        # involved in any oracle calculation, so disable it in the temporary copy.
        general_header = source_copy / "src" / "general.h"
        general_header.write_text(
            general_header.read_text().replace("#ifdef ENABLE_NLS", "#if 0", 1)
        )
        install_expression = (
            f"install.packages({str(source_copy)!r}, repos=NULL, type='source', "
            f"lib={str(library)!r}); "
            f"if (!requireNamespace('mgcv', lib.loc={str(library)!r}, "
            "quietly=TRUE)) quit(status=1)"
        )
        subprocess.run(["Rscript", "-e", install_expression], check=True)
        environment = os.environ.copy()
        environment["R_LIBS"] = str(library)
        subprocess.run(
            ["Rscript", str(HERE / "generate.R"), str(csv_output)],
            check=True,
            env=environment,
        )

        cases: dict[str, dict[str, object]] = {}
        for directory in sorted(path for path in csv_output.iterdir() if path.is_dir()):
            arrays = {
                "x": _matrix(directory / "x.csv"),
                "new_x": _matrix(directory / "new_x.csv"),
                "basis": _matrix(directory / "basis.csv"),
                "penalty": _matrix(directory / "penalty.csv"),
                "new_basis": _matrix(directory / "new_basis.csv"),
            }
            for name in (
                "transformed_basis",
                "transformed_penalty",
                "transformed_new_basis",
            ):
                path = directory / f"{name}.csv"
                if path.exists():
                    arrays[name] = _matrix(path)
            target = output / f"{directory.name}.npz"
            np.savez_compressed(target, **arrays)  # type: ignore[arg-type]  # ty: ignore[invalid-argument-type]
            cases[directory.name] = {
                **_metadata(directory / "metadata.txt"),
                "shapes": {key: list(value.shape) for key, value in arrays.items()},
                "sha256": hashlib.sha256(target.read_bytes()).hexdigest(),
            }

        commit = subprocess.run(
            ["git", "-C", str(source), "rev-parse", "HEAD"],
            check=True,
            capture_output=True,
            text=True,
        ).stdout.strip()
        manifest = {
            "mgcv_version": (csv_output / "version.txt").read_text().strip(),
            "mgcv_commit": commit,
            "generator": "generate.R",
            "cases": cases,
        }
        (output / "manifest.json").write_text(
            json.dumps(manifest, indent=2, sort_keys=True) + "\n"
        )


if __name__ == "__main__":
    main()

File: tests/mgcv_reference/generate.R
#!/usr/bin/env Rscript

# SPDX-License-Identifier: GPL-3.0-or-later
# Copyright (C) 2026 Johannes Brachem
#
# Generate mgcv oracle matrices.
# The adjacent generate.py driver installs the requested mgcv source tree into an
# isolated library, invokes this script, and converts these CSV intermediates to NPZ.

suppressPackageStartupMessages(library(mgcv))

args <- commandArgs(trailingOnly = TRUE)
if (length(args) != 1) stop("usage: generate.R OUTPUT_DIRECTORY")
output <- args[[1]]
dir.create(output, recursive = TRUE, showWarnings = FALSE)
writeLines(as.character(packageVersion("mgcv")), file.path(output, "version.txt"))
options(digits = 17)

write_matrix <- function(value, path) {
  write.table(
    data.matrix(value),
    file = path,
    row.names = FALSE,
    col.names = FALSE,
    sep = ","
  )
}

write_case <- function(
  name,
  specification,
  data,
  newdata,
  knots = NULL,
  include_transformed = FALSE
) {
  smooth <- smoothCon(
    specification,
    data = data,
    knots = knots,
    absorb.cons = FALSE,
    diagonal.penalty = FALSE,
    scale.penalty = FALSE
  )[[1]]
  directory <- file.path(output, name)
  dir.create(directory, showWarnings = FALSE)
  write_matrix(data, file.path(directory, "x.csv"))
  write_matrix(newdata, file.path(directory, "new_x.csv"))
  write_matrix(smooth$X, file.path(directory, "basis.csv"))
  write_matrix(smooth$S[[1]], file.path(directory, "penalty.csv"))
  write_matrix(PredictMat(smooth, newdata), file.path(directory, "new_basis.csv"))
  if (include_transformed) {
    transformed <- smoothCon(
      specification,
      data = data,
      knots = knots,
      absorb.cons = TRUE,
      diagonal.penalty = TRUE,
      scale.penalty = TRUE
    )[[1]]
    write_matrix(
      transformed$X,
      file.path(directory, "transformed_basis.csv")
    )
    write_matrix(
      transformed$S[[1]],
      file.path(directory, "transformed_penalty.csv")
    )
    write_matrix(
      PredictMat(transformed, newdata),
      file.path(directory, "transformed_new_basis.csv")
    )
  }
  writeLines(
    c(
      paste("rank", smooth$rank),
      paste("nullity", smooth$null.space.dim)
    ),
    file.path(directory, "metadata.txt")
  )
}

x <- sort(unique(c(seq(-1.8, 2.2, length.out = 31), -0.7, 0.15, 1.41)))
new_x <- c(-2.4, -1.8, -1.1, -0.05, 0.8, 1.75, 2.2, 2.8)
data_1d <- data.frame(x = x)
new_1d <- data.frame(x = new_x)

write_case(
  "ps",
  s(x, bs = "ps", k = 9, m = c(2, 2)),
  data_1d,
  new_1d,
  include_transformed = TRUE
)
write_case("bs", s(x, bs = "bs", k = 9, m = c(3, 2)), data_1d, new_1d)
write_case("cp", s(x, bs = "cp", k = 9, m = c(2, 2)), data_1d, new_1d)
write_case("cr", s(x, bs = "cr", k = 9), data_1d, new_1d)
write_case("cs", s(x, bs = "cs", k = 9), data_1d, new_1d)
write_case("cc", s(x, bs = "cc", k = 9), data_1d, new_1d)
write_case("tp_1d", s(x, bs = "tp", k = 9, m = 2), data_1d, new_1d)

x2 <- seq(-1.5, 1.7, length.out = 36)
y2 <- sin(seq(0.1, 4.9, length.out = 36)) + seq(-0.2, 0.25, length.out = 36)
data_2d <- data.frame(x = x2, y = y2)
new_2d <- data.frame(
  x = c(-1.7, -1.2, -0.3, 0.4, 1.1, 1.9),
  y = c(-1.1, 0.2, 0.95, -0.65, 0.35, 1.2)
)
write_case("tp_2d", s(x, y, bs = "tp", k = 12, m = 2), data_2d, new_2d)
write_case("ts_2d", s(x, y, bs = "ts", k = 12, m = 2), data_2d, new_2d)

write_case("gp_spherical", s(x, y, bs = "gp", k = 12, m = c(1, -1, 1)), data_2d, new_2d)
write_case("gp_power", s(x, y, bs = "gp", k = 12, m = c(2, -1, 1.4)), data_2d, new_2d)
write_case("gp_matern15", s(x, y, bs = "gp", k = 12, m = c(3, -1, 1)), data_2d, new_2d)
write_case("gp_matern25", s(x, y, bs = "gp", k = 12, m = c(4, 2.3, 1)), data_2d, new_2d)
write_case("gp_matern35_stationary", s(x, y, bs = "gp", k = 12, m = c(-5, -1, 1)), data_2d, new_2d)

regions <- factor(
  c("a", "b", "c", "d", "e", "f", "a", "c", "e", "f", "b", "d"),
  levels = c("a", "b", "c", "d", "e", "f")
)
new_regions <- factor(c("f", "a", "d", "c"), levels = levels(regions))
nb <- list(
  a = c("b"),
  b = c("a", "c"),
  c = c("b", "d"),
  d = c("c", "e"),
  e = c("d", "f"),
  f = c("e")
)
write_case(
  "mrf_full",
  s(region, bs = "mrf", k = -1, xt = list(nb = nb)),
  data.frame(region = regions),
  data.frame(region = new_regions)
)
write_case(
  "mrf_low_rank",
  s(region, bs = "mrf", k = 4, xt = list(nb = nb)),
  data.frame(region = regions),
  data.frame(region = new_regions)
)
